### PR TITLE
Introduce fix_remote_address_from_ip_trail_with_verification_key()

### PR DESCRIPTION
**Pull request by simonwheatley**
Work in progress

* [x] Add `fix_remote_address_from_ip_trail_with_verification_key()`
* [x] Abstract shared functionality
* [x] Tests for `fix_remote_address_from_ip_trail_with_verification_key()`
* [x] Tests for `is_valid_proxy_verification_key()`
* [x] Tests for `get_ip_addresses_from_ip_trail()`

### CODE REVIEW

**Comment by simonwheatley** [comment]
@mjangda Can I have an interim review to check my direction of travel, please? Questions: 

* Any issues with the abstraction into `get_ip_addresses_from_ip_trail()` and `is_valid_proxy_verification_key()`?
* The tests in `IP_Forward__Get_Ip_Addresses_From_Ip_Trail__Tests` significantly duplicate the tests in `IP_Forward_Tests`, also the tests in `IP_Forward__Is_Valid_Proxy_Verification_Key__Test` duplicate the tests in `IP_Forward__Fix_Remote_Address_With_Verification_Key__Test` (but there's less tests there)… should I favour tests in the utility type functions, e.g. `get_ip_addresses_from_ip_trail()`, and remove tests on the API functions, e.g. `fix_remote_address_from_ip_trail()` or just leave the duplication?

I'm aware I've got failing tests… will look tomorrow!

File: lib/proxy/ip-forward.php [hunk 1]
```diff
@@ -118,12 +122,111 @@ function fix_remote_address_with_verification_key( $user_ip, $submitted_verifica
...
+/**
+ * Validate the provided verification key against the one in config.
+ *
+ * @param string $submitted_verification_key The key to validate
+ *
+ * @return bool True if the key is valid
+ */
+function is_valid_proxy_verification_key( $submitted_verification_key ) {
+	$expected_verification_key = get_proxy_verification_key();
+	if ( ! hash_equals( $submitted_verification_key, $expected_verification_key ) ) {
```
Comment (mjangda) [inline lib/proxy/ip-forward.php]: Minor: we could simplify this as `return hash_equals( ... )`

File: tests/test-lib-proxy-ip-forward.php [hunk 2]
```diff
@@ -251,6 +329,29 @@ public function test__defined() {
 	}
 }
 
+class IP_Forward__Is_Valid_Proxy_Verification_Key__Test extends \PHPUnit_Framework_TestCase {
```
Comment (mjangda) [inline tests/test-lib-proxy-ip-forward.php]: For these to work, we'll either need to run each test in a separate process (see `IP_Forward__Get_Proxy_Verification_Key__Test` and define the constant for each) or mock the return value of `get_proxy_verification_key()`

File: lib/proxy/ip-forward.php (hunk 1 above)
Comment (simonwheatley) [inline lib/proxy/ip-forward.php]: Done: https://github.com/Automattic/vip-go-mu-plugins/pull/679/commits/223b0eff03a602e441b4a2bcaa4dc3e260d5c1b5

File: tests/test-lib-proxy-ip-forward.php (hunk 2 above)
Comment (simonwheatley) [inline tests/test-lib-proxy-ip-forward.php]: Done: https://github.com/Automattic/vip-go-mu-plugins/pull/679/commits/930027602a2049dbf7c298943f448f1d4d86ab29

**Comment by simonwheatley** [comment]
@mjangda Ready for your review again, please…

**Comment by mjangda** [comment]
🚢 🚢 🚢

**Comment by simonwheatley** [comment]
r95129-deploy
